core/src/main/scala/io/finch/Endpoint.scala

```diff
@@ -289,10 +289,12 @@ trait Endpoint[A] { self =>
    * @return an endpoint that will return the value of this reader if it is valid.
    *         Otherwise the future fails with an [[Error.NotValid]] error.
    */
-  final def should(rule: String)(predicate: A => Boolean): Endpoint[A] = mapAsync(a =>
-    if (predicate(a)) Future.value(a)
-    else Future.exception(Error.NotValid(self.item, "should " + rule))
-  )
+  final def should(rule: String)(predicate: A => Boolean): Endpoint[A] = {
+    validate { (a: A, item) =>
+      if (predicate(a)) Validated.Valid(a)
+      else Validated.Invalid(Errors.of(Error.NotValid(item, "should " + rule)))
```

Comment:
Besides the allocation that's also a change of semantics that I don't think we should do. Other than that I'd be happy to add the `validate` method. But I think to make it useful we also need to add a `Semigroup[Errors]` instance. It's unfortunate that it uses `NonEmptyList` though because lists don't support efficient concatenation.

Reply:
Ah right it would require a Semigroup for errors, but thats fairly straightforward to implement. This PR is fairly old (it was made prior to cats effect being introduced). I think it might make sense to close this and start fresh. In terms of performance ValidatedNec was not available when I wrote this, perhaps moving to that would address the performance concerns.